### PR TITLE
[12.0][ADD] sale_product_returnable

### PR DESCRIPTION
**Pull request by nikitavaghela**
#993

### CODE REVIEW

**Comment by nikitavaghela** [comment]
Hello @max3903, @dreispt ,
Please review this PR.

File: sale_product_returnable/models/__init__.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Copyright (C) 2019 Open Source Integrators
+# Copyright (C) 2019 Serpent Consulting Services Pvt. Ltd.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+from . import product_template
+# from . import stock_return_picking
```
Comment (bodedra) [inline sale_product_returnable/models/__init__.py]:
```suggestion

```

File: sale_product_returnable/README.rst [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+Do not contact contributors directly about support or help with technical issues.
+
+Credits
+=======
+
+Authors
+~~~~~~~
+
+* Open Source Integrators
+* Odoo Community Association (OCA)
```
Comment (bodedra) [inline sale_product_returnable/README.rst]:
```suggestion

```

File: sale_product_returnable/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    'name': 'Sale Product Retunable',
+    'version': '12.0.1.0.0',
+    'license': 'AGPL-3',
+    'category': 'Sales',
+    'author': 'Open Source Integrators, Odoo Community Association (OCA)',
+    'website': 'https://github.com/OCA/sale-workflow',
+    'development_status': 'Production/Stable',
+    'summary': """This module allows you to set some product as returnable
+                and create a reception when the sales order is confirmed.""",
+    'depends': ['sale_stock'],
```
Comment (bodedra) [inline sale_product_returnable/__manifest__.py:15]:
```suggestion
    'depends': ['sale_stock'],
    'maintainers': ['max3903'],
```

File: sale_product_returnable/models/sale_order.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# Copyright (C) 2019 Open Source Integrators
+# Copyright (C) 2019 Serpent Consulting Services Pvt. Ltd.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+
+from odoo import _, api, models
```
Comment (bodedra) [inline sale_product_returnable/models/sale_order.py]:
```suggestion
from odoo import api, models, _
```

**Review by bodedra — COMMENTED** [review]
Minor suggestions

File: sale_product_returnable/models/__init__.py (hunk 1 above)
Comment (nikitavaghela) [reply sale_product_returnable/models/__init__.py]: Done

**Comment by max3903** [comment]
@nikitavaghela Please review and merge https://github.com/nikitavaghela/sale-workflow/pull/1

File: sale_product_returnable/tests/test_sale_product_returnable.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                'product_id': self.product_10.id,
+                'product_uom_qty': 1,
+                'product_uom': self.product_uom_unit.id,
+                'price_unit': 750.00,
+                'tax_id': [(4, self.percent_tax.id, 0)]
+            })],
+        })
+
+        self.sale.action_confirm()
+        self.assertTrue(self.sale.state == 'sale')
```
Comment (max3903) [inline sale_product_returnable/tests/test_sale_product_returnable.py:54]: @nikitavaghela Can you also check that you get a receipt for the returned product with the same quantity?

Comment (nikitavaghela) [reply sale_product_returnable/tests/test_sale_product_returnable.py:54]: No , I will add that.

**Comment by nikitavaghela** [comment]
@max3903  Added test case for the compared quantity of picking and return picking. please review.

**Review by max3903 — APPROVED** [review]

File: sale_product_returnable/models/sale_order.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            'product_uom': sale_order_line.product_id.uom_id.id,
+            'origin': self.name,
+            'group_id': self.procurement_group_id.id,
+        }
+
+    @api.multi
+    def action_confirm(self):
+        """Action Confirm.
+        Create return pickings when confirming sale orders.
+        """
```
Comment (bodedra) [inline sale_product_returnable/models/sale_order.py:40]:
```suggestion
        """
        res = super(SaleOrder, order).action_confirm()
```

File: sale_product_returnable/models/sale_order.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        for order in self:
+            res = super(SaleOrder, order).action_confirm()
```
Comment (bodedra) [inline sale_product_returnable/models/sale_order.py]:
```suggestion

```

File: sale_product_returnable/models/sale_order.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+                    new_picking = \
+                        self.env['stock.picking'].create(new_picking_vals)
+                    new_picking.message_post_with_view(
+                        'mail.message_origin_link',
+                        values={
+                            'self': new_picking, 'origin': self},
+                        subtype_id=self.env.ref('mail.mt_note').id)
+                    new_picking.action_confirm()
+                    new_picking.action_assign()
+            return res
```
Comment (bodedra) [inline sale_product_returnable/models/sale_order.py]:
```suggestion
        return res
```

**Review by bodedra — CHANGES_REQUESTED** [review]
Multi sale order confirmation event will be failed with this module. Please review suggesting changes.

File: sale_product_returnable/models/sale_order.py (hunk 6 above)
Comment (nikitavaghela) [reply sale_product_returnable/models/sale_order.py:40]: Done

**Review by nelsonramirezs — APPROVED** [review]

**Review by bodedra — APPROVED** [review]
@max3903 @nikitavaghela I have completed functional testing at http://3403383-1019-302842.runbot1-2.odoo-community.org/web?debug=1#id=25&action=405&model=sale.order&view_type=form&menu_id=254 and LGTM.

It would be nice if we link return picking (incoming shipment) to sale order. So we get correct delivery order counter in sale order form view.

To achieve above, we need to set 'sale_id' value in return picking.

**Comment by nikitavaghela** [comment]
> @max3903 @nikitavaghela I have completed functional testing at http://3403383-1019-302842.runbot1-2.odoo-community.org/web?debug=1#id=25&action=405&model=sale.order&view_type=form&menu_id=254 and LGTM.
> 
> It would be nice if we link return picking (incoming shipment) to sale order. So we get correct delivery order counter in sale order form view.
> 
> To achieve above, we need to set 'sale_id' value in return picking.

In stock_picking `sale_id` field is a related field of procurement group (`group_id`) so sale order and picking relation are already linked by sale_id field.

**Comment by bodedra** [comment]
Okay. May you check why it's not increase counter in sale order.

Firefox_Screenshot_2019-12-31T08-03-41 244Z

Here is two test case:

1. Create a sale order with return product checked, confirm it and verify delivery count.
2. Create a sale order without return product checked, confirm it, validate picking, return it and verify delivery count. Now you will see 2, 1 delivery order and 1 incoming shipment.

**Comment by nikitavaghela** [comment]
> Okay. May you check why it's not increase counter in sale order.
> 
> Firefox_Screenshot_2019-12-31T08-03-41 244Z
> 
> Here is two test case:
> 
>     1. Create a sale order with return product checked, confirm it and verify delivery count.
> 
>     2. Create a sale order without return product checked, confirm it and verify delivery count.

As per #993 requirement, created return picking if the sale order line contains the product is returnable True. so if a product is not returnable then return picking is not created. so the delivery counter is not increased.

**Comment by bodedra** [comment]
I have updated test case in my previous comment.

**Comment by nikitavaghela** [comment]
> Okay. May you check why it's not increase counter in sale order.
> 
> Firefox_Screenshot_2019-12-31T08-03-41 244Z
> 
> Here is two test case:
> 
>     1. Create a sale order with return product checked, confirm it and verify delivery count.
> 
>     2. Create a sale order without return product checked, confirm it and verify delivery count.

The delivery count is not increased because of the sale_procurement_group_by_line module. In this module changed the procurement group configuration in the sale order. sale_id is the related field of the procurement group so we can not update it.
Cc @max3903

**Pull request merged**